### PR TITLE
[dynamo] Forward fix a bunch of distributed collective allow fixes

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111171

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/functions.py [hunk 1]
```diff
@@ -551,7 +551,7 @@ def _traceable_collectives_source(fn):
     assert fn in valid_values
     inner_name = fn.__name__
     path_source = AttrSource(
-        base=AttrSource(base=GlobalSource(global_name="torch"), member="distributed"),
```
Comment (voznesenskym) [inline torch/_dynamo/variables/functions.py:554]: this never worked - it was just unused. The moment we took a guard on it, it all fell apart. `__import` is the dynamo prefix for modules in the global name. Most module go through this munging, so we need to simulate it to keep guards alright.

File: torch/_dynamo/allowed_functions.py [hunk 2]
```diff
@@ -237,13 +237,14 @@ def _find_torch_objects(module):
...
     if config.trace_distributed:
+        from torch.distributed import _functional_collectives_impl as fci
         for f in [
-            torch.distributed._functional_collectives_impl._all_gather_into_tensor,
-            torch.distributed._functional_collectives_impl._all_reduce,
-            torch.distributed._functional_collectives_impl._reduce_scatter_tensor,
-            torch.distributed._functional_collectives_impl._all_reduce_coalesced,
-            torch.distributed._functional_collectives_impl._all_gather_into_tensor_coalesced,
-            torch.distributed._functional_collectives_impl._reduce_scatter_tensor_coalesced,
+            fci._all_gather_into_tensor,
```
Comment (voznesenskym) [inline torch/_dynamo/allowed_functions.py:242]: I was wrong @ezyang, mea culpa

File: torch/_dynamo/skipfiles.py [hunk 3]
```diff
@@ -228,10 +222,12 @@ def _module_dir(m: types.ModuleType):
     "torch.testing",
     "torch.utils._contextlib",
     "torch.utils._pytree",
+    "torch.distributed",
```
Comment (voznesenskym) [inline torch/_dynamo/skipfiles.py:225]: For @yanboliang

Comment (yanboliang) [reply torch/_dynamo/skipfiles.py:225]: Thanks for fixing this! I think you have to remove L225, otherwise, the MacOS tests would be failed.

**Review by yanboliang — APPROVED** [review]

File: torch/_dynamo/skipfiles.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/skipfiles.py:225]: oh, yes, gc gc

**Pull request closed without merging**